src/vs/gitpod/node/gitpodInsightsAppender.ts

```diff
@@ -0,0 +1,181 @@
+	const instanceId = workspaceInfo.getInstanceId();
+
+	const getTokenRequest = new GetTokenRequest();
+	getTokenRequest.setKind('gitpod');
+	getTokenRequest.setHost(gitpodApiHost);
+	getTokenRequest.addScope('function:trackEvent');
+
+	const getTokenResponse = await util.promisify(supervisor.token.getToken.bind(supervisor.token, getTokenRequest, supervisor.metadata, {
+		deadline: Date.now() + supervisor.deadlines.long
+	}))();
+	const serverToken = getTokenResponse.getToken();
+
+	return {
+		serverToken,
+		gitpodHost,
+		gitpodApiEndpoint,
+		workspaceId,
+		instanceId
+	};
+}
+
+async function getClient(serverToken: string, gitpodHost: string, gitpodApiEndpoint: string): Promise<GitpodConnection> {
+	const factory = new JsonRpcProxyFactory<GitpodServer>();
+	const gitpodService = new GitpodServiceImpl<GitpodClient, GitpodServer>(factory.createProxy()) as GitpodConnection;
+
+	const webSocket = new (ReconnectingWebSocket as any)(gitpodApiEndpoint, undefined, {
+		maxReconnectionDelay: 10000,
+		minReconnectionDelay: 1000,
+		reconnectionDelayGrowFactor: 1.3,
+		connectionTimeout: 10000,
+		maxRetries: Infinity,
+		debug: false,
+		startClosed: false,
+		WebSocket: class extends WebSocket {
+			constructor(address: string, protocols?: string | string[]) {
+				super(address, protocols, {
+					headers: {
+						'Origin': new URL(gitpodHost).origin,
+						'Authorization': `Bearer ${serverToken}`
+					}
+				});
+			}
+		}
+	});
+	webSocket.onerror = console.error;
+	doListen({
+		webSocket: webSocket as any,
+		onConnection: connection => factory.listen(connection),
+		logger: new ConsoleLogger()
+	});
+
+	return gitpodService;
+}
+
+const formatEventName = (str: string) => {
+	return str
+		.replace(/^[A-Z]/g, letter => letter.toLowerCase())
+		.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`)
+		.replace(/[^\w]/g, '_');
+};
+
+export class GitpodInsightsAppender implements ITelemetryAppender {
+
+	private _asyncAIClient: Promise<GitpodConnection> | null;
+	private _defaultData: { [key: string]: any } = Object.create(null);
+
+	constructor() {
+		this._asyncAIClient = null;
+	}
+
+	private _withAIClient(callback: (aiClient: Pick<GitpodServer, 'trackEvent'>) => void): void {
+		if (!this._asyncAIClient) {
+			this._asyncAIClient = getSupervisorData().then(
+				(supervisorData) => {
+					this._defaultData['workspaceId'] = supervisorData.workspaceId;
+					this._defaultData['instanceId'] = supervisorData.instanceId;
+
+					return getClient(supervisorData.serverToken, supervisorData.gitpodHost, supervisorData.gitpodApiEndpoint);
+				}
+			);
+		}
+
+		this._asyncAIClient.then(
+			(aiClient) => {
+				callback(aiClient.server);
+			},
+			(err) => {
+				onUnexpectedError(err);
+				console.error(err);
+			}
+		);
+	}
+
+	log(eventName: string, data?: any): void {
+		data = mixin(data, this._defaultData);
+		data = validateTelemetryData(data);
+
+		if (eventName !== 'workbenchActionExecuted') {
```

Comment:
@jeanp413 let's start but tracking first read and write interactions, so we don't product many data and addressing https://github.com/gitpod-io/gitpod/issues/6981

We can have following event:
```js
{
    eventName: 'vscode_file_access`,
    properties: {
        // first `editorOpened` or `fileGet` fires `read`
        // first `filePut` fires `write`
        kind: 'read' | 'write',
        // from supervisor info endpoint
        workspaceId: string,
        // from supervisor info endpoint
        workspaceInstanceId: string,
        // from MS telemetry event
        sessionId: string,
        // from MS telemetry event
        timestamp: Date
    }
}
```

It is important that we care about only first to avoid spamming segment.

We need also kind of a function which does conversion from MS telemetry event to our event similar how @loujaybee is suggested in https://github.com/gitpod-io/openvscode-server/pull/257#issuecomment-994405752 It should help later adding more events for other team members.

It will be all for the scope of PR. After that we can work one by one with concrete issues.

Reply:
I added the logic to send just the first events but they are misleading as they are triggered when the getting started page is shown, also when the gitpod.yml file is read from the extension

Reply:
Is it because `editorOpened` or `fileGet`? 
Maybe we can filter down `editorOpened` for files with `file` scheme?

Reply:
Will take a look but not sure if that is gonna be reliable enough. either way I think this PR is done and we can polish this event in following PRs related to each gitpod telemetry issue

Reply:
I ended up creating separate event for `editorOpened` and filtering by `'workbench.editors.files.fileEditorInput'`
`fileGet` gets fired automatically by the gitpod-web extension when reading the `gitpod.yml` file.
If this needs more tweaking lets address them in following PRs

Reply:
yeah it sounds like `fileGet` and `filePut` more track fs access by any source (i.e. api and user triggered), we should not rely on it.

Reply:
so should I delete them? and just fire `editorOpened`?